cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -194,6 +214,47 @@ protected void run() {
       return;
     }
 
+    if (instanceId == 0) {
+      // todo iterate migration resolution order list and schedule execution based on data availability.
+
+      // transfer aggregates metrics data at the beginning, in a single run to avoid conflicts with new metrics written
+      // as metrics table is non-transactional. having conflict will provide incorrect results.
+      // it is okay to run this without a limit on records, as the total records in aggregates table will be equal
+      // to the total unique metrics in the system.
+      // we will schedule the transfer of metrics data for time resolution tables
+      // in a single thread executor with limit on records to transfer.
+      // since the smaller resolution tables have timestamp information in the rowkey, they wont have conflict for most
+      // of their data, except for the few metrics emitted in the last hour/minute.
+
+      if (!skipMigration) {
+        DataMigrator dataMigrator = new DataMigrator(metricDatasetFactory, cConfiguration, hConf, hBaseTableUtil,
+                                                     ImmutableList.of(Integer.MAX_VALUE), Integer.MAX_VALUE);
+        dataMigrator.run();
+
+        // migrate 1hr resolution table first, then 1min resolution table, limit 1000 records to scan/transfer in a run.
+        dataMigrator = new DataMigrator(metricDatasetFactory, cConfiguration, hConf, hBaseTableUtil,
+                                        ImmutableList.of(3600, 60), 1000);
+
+        ScheduledExecutorService scheduledExecutorService =
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-data-migrator"));
+
+        // 5-min initial delay and scheduled to run each min after that
+        scheduledExecutorService.scheduleAtFixedRate(dataMigrator, 5, 1, TimeUnit.MINUTES);
+        LOG.info("Scheduled metrics migration thread for resolution 3600, 60 tables");
+
+        MetricsTableDeleter tableDeletion =
+          new MetricsTableDeleter(metricDatasetFactory, cConfiguration, hConf, hBaseTableUtil,
+                                  ImmutableList.of(Integer.MAX_VALUE, 3600, 60, 1));
+        ScheduledExecutorService scheduledExecutorServiceDelete =
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-table-deleter"));
+
+        // scheduled to run every 2 hours
+        scheduledExecutorServiceDelete.scheduleAtFixedRate(tableDeletion, 1, 2, TimeUnit.HOURS);
```

Comment:
why is the deletion in a separate thread? Can't the dataMigrator delete the table when it has migrated all data?